### ISSUE TITLE
Question about U-net architecture for segmentation

### ISSUE THREAD

**Issue by joriswuts** [Project-MONAI/MONAI#464]
Hi, 

I am a master student biomedical engineering at the University of Ghent. I have been using MONAI for my master thesis the last couple of months which is about the segmentation of bone lesions from multi-modal whole body MRI.  When comparing the model architecture of MONAI with most U-nets that are described in literature I noticed that the MONAI U-net used strided convolutions instead of max-pooling for downsampling. I was wondering why that is and if there is are any  reference papers suggesting to do this. 

Kind regards,
Joris Wuts

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Could you please help provide more information here as you initialized the UNet code?
Thanks.

**Comment by joriswuts** [comment]
Hi,

I initialized the UNet with this code: 
monai.networks.nets. model = UNet(dimensions=3, in_channels=3, out_channels=2, channels=(16, 32, 64,128,256), strides=(2, 2,2,2), num_res_units=2, norm=Norm.BATCH).to(device)
The model works great for the purpose I am using it but for the report of my master's dissertation I wanted to give an in depth visualisation of the model. 

When visualising the computational graph in Tensorboard I noticed there are no max pooling layers and that the downsampling is done by the strided convolutions. 

Also on the upsampling part I am a bit confused as to where the residual layers are and how they connect. It seems to me that is it kind a similar to the model below. Is there a schematic visualisation of the model somewhere? That would really help me to understand the model better.
Thanks in advance, 

Joris Wuts 
  ( https://raw.githubusercontent.com/mattmacy/vnet.pytorch/master/images/diagram.png)

**Comment by ericspod** [comment]
Hi @joriswuts,
It might be helpful to print the network to stdout, it prints a tree structure which is a bit hard to read at first but does show the structure of the UNet correctly. Upsample layers are given as:

```
(2): Sequential(
  (conv): Convolution(
	(conv): ConvTranspose3d(64, 16, kernel_size=(3, 3, 3), stride=(2, 2, 2), padding=(1, 1, 1), output_padding=(1, 1, 1))
	(norm): BatchNorm3d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
	(act): PReLU(num_parameters=1)
  )
  (resunit): ResidualUnit(
	(conv): Sequential(
	  (unit0): Convolution(
		(conv): Conv3d(16, 16, kernel_size=(3, 3, 3), stride=(1, 1, 1), padding=(1, 1, 1))
		(norm): BatchNorm3d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
		(act): PReLU(num_parameters=1)
	  )
	)
	(residual): Identity()
  )
)
```

What this boils down to is a sequence of layers comprised of transpose convolution, normalization, activation, followed by a residual unit with a single sequence of convolution-normalization-activation layers.

A way of thinking about UNet is that it is a tower of large layers containing one level of the encode and decode paths with the skip connection between them. You can then describe the structure of the layer and then the whole network as a stack of these. I've attached a diagram of this I've used in other papers, it's very similar to your configuration other than the use of instance norm. 

![unet](https://user-images.githubusercontent.com/17726042/83431822-dd36dd80-a42f-11ea-8cdb-c2994cefd9f8.png)

**Comment by joriswuts** [comment]
@ericspod  
Thanks a lot for this information. I was just searching your other papers and found a clarifying explanation about the usage of strided convolutions instead of max pooling layers.  

Kind regards,
Joris Wuts

**Comment by joriswuts** [comment]
@ericspod, Just to be completely sure. Could it be that the residual connection in my model have a stride one convolution in them instead of a simple identity operation. I added an image of the first downsampling block in tensorboard. I noticed that a stride 1 convolution in the residual layer is also proposed in several papers that got referenced in the documents of DLTK and MONAI. (e.g. https://arxiv.org/pdf/1603.05027.pdf) 
If I interpret the code of MONAI correctly the convolution is also in the residual part of the upsampling path. 
![Schermafbeelding 2020-06-01 om 21 24 03](https://user-images.githubusercontent.com/63357811/83446114-77a61980-a44e-11ea-8798-c6bbcaaa02d9.png)

**Comment by ericspod** [comment]
Residual units are implemented in UNet with the [ResidualUnit](https://github.com/Project-MONAI/MONAI/blob/7cd65614da81eeff261a14abdf18bd07a20abfcc/monai/networks/blocks/convolutions.py#L81) class, the residual part uses a convolution to change the [input dimensions to match the output dimensions if this is necessary](https://github.com/Project-MONAI/MONAI/blob/7cd65614da81eeff261a14abdf18bd07a20abfcc/monai/networks/blocks/convolutions.py#L141) but will use `nn.Identity` if not. This way you don't have a needless convolutional layer in the residual path. In [the upsample stream of the network](https://github.com/Project-MONAI/MONAI/blob/7cd65614da81eeff261a14abdf18bd07a20abfcc/monai/networks/nets/unet.py#L96) the residual unit will come after the transpose convolution so it won't need to do this. In the [downsampling part of the network](https://github.com/Project-MONAI/MONAI/blob/7cd65614da81eeff261a14abdf18bd07a20abfcc/monai/networks/nets/unet.py#L77) the residual path will have a convolution since the number of channels and dimensions do change.

**Comment by kissievendor** [comment]
@ericspod I try to recreate what you show here.  I print the model using summary (by torchsummary).
The downsampling is the same as in your example. However, the upsampling part seems to be different. Can you clarrify this for me? 

![image](https://user-images.githubusercontent.com/54887561/99657944-1a833c00-2a5f-11eb-8f02-c2b0dde7a0a8.png)


> Hi @joriswuts,
> It might be helpful to print the network to stdout, it prints a tree structure which is a bit hard to read at first but does show the structure of the UNet correctly. Upsample layers are given as:
> 
> ```
> (2): Sequential(
>   (conv): Convolution(
> 	(conv): ConvTranspose3d(64, 16, kernel_size=(3, 3, 3), stride=(2, 2, 2), padding=(1, 1, 1), output_padding=(1, 1, 1))
> 	(norm): BatchNorm3d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> 	(act): PReLU(num_parameters=1)
>   )
>   (resunit): ResidualUnit(
> 	(conv): Sequential(
> 	  (unit0): Convolution(
> 		(conv): Conv3d(16, 16, kernel_size=(3, 3, 3), stride=(1, 1, 1), padding=(1, 1, 1))
> 		(norm): BatchNorm3d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> 		(act): PReLU(num_parameters=1)
> 	  )
> 	)
> 	(residual): Identity()
>   )
> )
> ```
> 
> What this boils down to is a sequence of layers comprised of transpose convolution, normalization, activation, followed by a residual unit with a single sequence of convolution-normalization-activation layers.
> 
> A way of thinking about UNet is that it is a tower of large layers containing one level of the encode and decode paths with the skip connection between them. You can then describe the structure of the layer and then the whole network as a stack of these. I've attached a diagram of this I've used in other papers, it's very similar to your configuration other than the use of instance norm.
> 
> ![unet](https://user-images.githubusercontent.com/17726042/83431822-dd36dd80-a42f-11ea-8cdb-c2994cefd9f8.png)

**Comment by AnnaKlemen** [comment]
@ericspod Is it possible to find the see the settings used in the references paper (Left-Ventricle Quantification Using Residual U-Net) e.g. optimizer, learning rate, weight decay, learning rate scheduler ect. ? :) 
I have tried with the setting from the "Road Extraction by Deep Residual U-Net"-paper, but it really does not work on my problem. (SGD, lr start: 0.001, reduced by 0.1 every 20 epochs)

**Comment by ericspod** [comment]
I used Adam with a learning rate of 0.001 with default values otherwise. I didn't do learning rate scheduling in this paper.

**Comment by AnnaKlemen** [comment]
> I used Adam with a learning rate of 0.001 with default values otherwise. I didn't do learning rate scheduling in this paper.

Thanks,  that is also what I have most luck with using :D

**Comment by r00tdotexe** [comment]
> Hi,
> 
> I am a master student biomedical engineering at the University of Ghent. I have been using MONAI for my master thesis the last couple of months which is about the segmentation of bone lesions from multi-modal whole body MRI. When comparing the model architecture of MONAI with most U-nets that are described in literature I noticed that the MONAI U-net used strided convolutions instead of max-pooling for downsampling. I was wondering why that is and if there is are any reference papers suggesting to do this.
> 
> Kind regards, Joris Wuts

Hey there,

I am currently in the same spot as you, writing a thesis for my biomed. engineering studies. I am using the MONAI U-Net, too, and yes, it seems, the max pooling operations are replaced by strided convolutions. I cam to the same conclusion by printing out the model sequences.
